dataflow/conftest.py

```diff
@@ -0,0 +1,378 @@
+            project, f"{subscription_name}-{UUID}"
+        )
+        subscription = subscriber.create_subscription(subscription_path, topic_path)
+
+        print(f"pubsub_subscription: {subscription.name}")
+        yield subscription.name
+
+        # Due to the pinned library dependencies in apache-beam, client
+        # library throws an error upon deletion.
+        # We use gcloud for a workaround. See also:
+        # https://github.com/GoogleCloudPlatform/python-docs-samples/issues/4492
+        cmd = [
+            "gcloud",
+            "pubsub",
+            "--project",
+            project,
+            "subscriptions",
+            "delete",
+            subscription.name,
+        ]
+        print(cmd)
+        subprocess.run(cmd, check=True)
+
+    @staticmethod
+    def pubsub_publisher(
+        topic_path: str,
+        new_msg: Callable[[int], str] = lambda i: json.dumps(
+            {"id": i, "content": f"message {i}"}
+        ),
+        sleep_sec: int = 1,
+    ) -> bool:
+        from google.cloud import pubsub
+
+        def _infinite_publish_job() -> None:
+            publisher_client = pubsub.PublisherClient()
+            for i in itertools.count():
+                msg = new_msg(i)
+                publisher_client.publish(topic_path, msg.encode("utf-8")).result()
+                time.sleep(sleep_sec)
+
+        # Start a subprocess in the background to do the publishing.
+        print(f"Starting publisher on {topic_path}")
+        p = mp.Process(target=_infinite_publish_job)
+        p.start()
+
+        yield p.is_alive()
+
+        # For cleanup, terminate the background process.
+        print("Stopping publisher")
+        p.join(timeout=0)
+        p.terminate()
+
+    @staticmethod
+    def container_image(
+        image_path: str,
+        project: str = PROJECT,
+        tag: str = "latest",
+    ) -> str:
+        image_name = f"gcr.io/{project}/{image_path}-{UUID}:{tag}"
+        cmd = ["gcloud", "auth", "configure-docker"]
+        print(cmd)
+        subprocess.run(cmd, check=True)
+        cmd = [
+            "gcloud",
+            "builds",
+            "submit",
+            f"--project={project}",
+            f"--tag={image_name}",
+            ".",
+        ]
+        print(cmd)
+        subprocess.run(cmd, check=True)
+
+        print(f"container_image: {image_name}")
+        yield image_name
+
+        cmd = [
+            "gcloud",
+            "container",
+            "images",
+            "delete",
+            image_name,
+            f"--project={project}",
+            "--quiet",
+        ]
+        print(cmd)
+        subprocess.run(cmd, check=True)
+
+    @staticmethod
+    def dataflow_job_id_from_job_name(
+        job_name: str,
+        project: str = PROJECT,
+    ) -> Optional[str]:
+        from googleapiclient.discovery import build
+
+        dataflow = build("dataflow", "v1b3")
+
+        # Only return the 50 most recent results - our job is likely to be in here.
```

Comment:
If the job is not found in the 50 most recent results and we need to increase it, should this be done manually?

Reply:
I think most of the time it will be within the latest 50 jobs. It's only active jobs, so any finished job doesn't count. In the unlikely case it's not there due to a spike of jobs at that time, re-running the tests should fix it. I think it's okay to have it like this for now, and if we see it's causing flaky tests, we can increase it in the future.

As a side note, I was using this function before, but now I'm getting the job ID directly from the command output, so none of the tests are actually using this function. I thought it could be useful in the future, so I left it there, but we won't be having any flaky tests due to this for now :)

I'm also happy to remove it if we prefer to clean this.

Reply:
Sounds good. Could you file an issue to track this?

Reply:
Opened #6094 to track this